src/DurableTask.AzureStorage/TimeoutHandler.cs

```diff
@@ -72,9 +77,32 @@ internal static class TimeoutHandler
 
                             // Delay to ensure the ETW event gets written
                             await Task.Delay(TimeSpan.FromSeconds(3));
-                            Environment.FailFast(message);
 
-                            // Should never be hit, due to above FailFast() call.
+                            bool executeFailFast = true;
+                            Task<bool> gracefulShutdownTask = Task.Run(async () =>
+                            {
+                                try
+                                {
+                                    return await settings.OnImminentFailFast(message);
+                                }
+                                catch (Exception)
+                                {
+                                    return true;
+                                }
+                            });
+
+                            await Task.WhenAny(gracefulShutdownTask, Task.Delay(TimeSpan.FromSeconds(35)));
+
+                            if (gracefulShutdownTask.IsCompleted)
+                            {
+                                executeFailFast = gracefulShutdownTask.Result;
+                            }
+
+                            if (executeFailFast)
+                            {
+                                TimeoutHandler.ProcessKillAction(message);
+                            }
+
                             return default(T);
```

Comment:
I appreciate these updates. One final thing - since we may not always FailFast, instead of returning `default(T)`, can we instead throw a TimeoutException?
```suggestion
                            throw new TimeoutException(message);
```

Reply:
Done.